hackathon_site/hardware/serializers.py

```diff
@@ -46,9 +54,128 @@ def get_unique_hardware_count(obj: Category):
+
+
+class OrderCreateSerializer(serializers.Serializer):
+    class OrderCreateHardwareSerializer(serializers.Serializer):
+        id = serializers.PrimaryKeyRelatedField(
+            queryset=Hardware.objects.all(), many=False, required=True
+        )
+        quantity = serializers.IntegerField(required=True)
+
+    hardware = OrderCreateHardwareSerializer(many=True, required=True)
+
+    @staticmethod
+    def merge_requests(hardware_requests):
+        return functools.reduce(
+            lambda x, y: x + y,
+            [Counter({e["id"]: e["quantity"]}) for e in hardware_requests],
+        )
+
+    # check that the requests are within per-team constraints
+    def validate(self, data):
+        requested_hardware = self.merge_requests(hardware_requests=data["hardware"])
+        hardware_query = (
+            Hardware.objects.filter(
+                id__in=[hardware.id for hardware in requested_hardware.keys()],
+            )
+            .all()
+            .prefetch_related("categories", "order_items")
+        )
+        team_unreturned_orders = hardware_query.annotate(
+            past_order_count=Count(
+                "order_items",
+                filter=Q(order_items__part_returned_health__isnull=True)
+                & ~Q(order_items__order__status="Cancelled")
+                & Q(order_items__order__team=self.context["request"].user.profile.team),
+            )
+        )
+        category_counts = dict()
+        error_messages = []
+        for (hardware, requested_quantity) in requested_hardware.items():
+            try:
+                team_hardware = team_unreturned_orders.get(id=hardware.id)
+            except ObjectDoesNotExist:
+                team_hardware = None
+            team_hardware_count = getattr(team_hardware, "past_order_count", 0)
+            if hardware.max_per_team < (team_hardware_count + requested_quantity):
+                error_messages.append("Hardware {} limit reached".format(hardware.name))
+            for category in hardware.categories.all():
+                category_counts[category] = (
+                    category_counts.get(category, 0) + team_hardware_count
+                )
+        for (category, count) in category_counts.items():
+            if category.max_per_team < count:
+                error_messages.append("Category {} limit reached".format(category.name))
+        if error_messages:
+            error_message = "; ".join(error_messages)
+            raise serializers.ValidationError(error_message)
+        return data
+
+    def create(self, validated_data):
+        requested_hardware = self.merge_requests(
+            hardware_requests=validated_data["hardware"]
+        )
+        relevant_past_order_items = (
+            OrderItem.objects.filter(hardware__in=requested_hardware.keys(),)
+            .exclude(order__status="Cancelled")
+            .select_related("hardware")
+        )
+        team_past_hardware_counts = Counter(
```

Comment:
Can you convert this to pure database logic like you did above? My tiny python brain breaks as soon as map is involved (and doing it in the database is more efficient)

Reply:
updated